src/ripple/shamap/impl/SHAMapInnerNode.cpp

```diff
@@ -33,26 +33,718 @@
+            boost::fast_pool_allocator_tag,
+            arrayChunkSizeBytes[I],
+            boost::default_user_allocator_new_delete,
+            std::mutex,
+            chunksPerBlock[I],
+            chunksPerBlock[I]>::is_from...,
+    };
+}
+std::array<std::function<bool(void*)>, boundaries.size()> const
+    isFromArrayFuns =
+        initIsFromArrayFuns(std::make_index_sequence<boundaries.size()>{});
+
+// This function returns an untagged pointer
+[[nodiscard]] inline std::pair<std::uint8_t, void*>
+allocateArrays(std::uint8_t numChildren)
+{
+    auto const i = boundariesIndex(numChildren);
+    return {i, allocateArrayFuns[i]()};
+}
+
+// This function takes an untagged pointer
+inline void
+deallocateArrays(std::uint8_t boundaryIndex, void* p)
+{
+    assert(isFromArrayFuns[boundaryIndex](p));
+    freeArrayFuns[boundaryIndex](p);
+}
+#else
+
+template <std::size_t... I>
+std::array<std::pmr::synchronized_pool_resource, boundaries.size()>
+    initPmrArrayFuns(std::index_sequence<I...>)
+{
+    return std::array<std::pmr::synchronized_pool_resource, boundaries.size()>{
+        std::pmr::synchronized_pool_resource{std::pmr::pool_options{
+            .max_blocks_per_chunk = chunksPerBlock[I],
+            .largest_required_pool_block = chunksPerBlock[I]}}...,
+    };
+}
+std::array<std::pmr::synchronized_pool_resource, boundaries.size()>
+    pmrArrayFuns =
+        initPmrArrayFuns(std::make_index_sequence<boundaries.size()>{});
+
+// This function returns an untagged pointer
+[[nodiscard]] inline std::pair<std::uint8_t, void*>
+allocateArrays(std::uint8_t numChildren)
+{
+    auto const i = boundariesIndex(numChildren);
+    return {i, pmrArrayFuns[i].allocate(arrayChunkSizeBytes[i])};
+}
+
+// This function takes an untagged pointer
+inline void
+deallocateArrays(std::uint8_t boundaryIndex, void* p)
+{
+    return pmrArrayFuns[boundaryIndex].deallocate(
+        p, arrayChunkSizeBytes[boundaryIndex]);
+}
+#endif
+
+[[nodiscard]] inline int
+popcnt16(std::uint16_t a)
+{
+#if defined(__clang__) || defined(__GNUC__)
+    return __builtin_popcount(a);
+#else
+    // fallback to table lookup
+    static auto constexpr const tbl = []() {
+        std::array<std::uint8_t, 256> ret{};
+        for (int i = 0; i != 256; ++i)
+        {
+            for (int j = 0; j != 8; ++j)
+            {
+                if (i & (1 << j))
+                    ret[i]++;
+            }
+        }
+        return ret;
+    }();
+    return tbl[v & 0xff] + tbl[v >> 8];
+#endif
+}
+
+}  // namespace
+
+template <class F>
+void
+SHAMapInnerNode::TaggedPointer::iterChildren(std::uint16_t isBranch, F&& f)
+    const
+{
+    auto [numAllocated, hashes, _] = getHashesAndChildren();
+    if (numAllocated == SHAMapInnerNode::branchFactor)
+    {
+        // dense case
+        for (int i = 0; i < SHAMapInnerNode::branchFactor; ++i)
+            f(hashes[i]);
+    }
+    else
+    {
+        // sparse case
+        int curHashI = 0;
+        SHAMapHash zero;
```

Comment:
Should be `static const`; in fact, since this is a template you should consider making this a `static const` member of `SHAMapInnerNode`

Reply:
Good idea - although I just made it a file-level variable instead of a static member of `SHAMapInnerNode`. Fixed in `2fa4a8d59e [fold] Doxygen comments and misc cleanups 1:`